File: packages/pdp-geo-generator-agent/src/keyword-normalizer.ts
import type {
  PdpGeoEvidence,
  PdpGeoGeneratorOptions,
  PdpGeoKeywordCorrection,
  PdpGeoKeywordNormalizationRequest,
  PdpGeoKeywordNormalizationResult,
  PdpGeoKeywordNormalizationSettings,
  PdpGeoKeywordNormalizer,
  PdpGeoLocale,
  PdpGeoTokenUsage,
  PdpProductSignal
} from "./types";
import { temperatureBody } from "./copy-refiner";

interface KeywordNormalizationApplication {
  product: PdpProductSignal;
  evidence: PdpGeoEvidence[];
  warnings: string[];
  usage?: PdpGeoTokenUsage;
}

interface ModelBackedKeywordNormalizerConfig {
  provider: Exclude<PdpGeoGeneratorOptions["provider"], undefined>;
  apiKey?: string;
  model?: string;
  endpoint?: string;
  deployment?: string;
  apiVersion?: string;
  temperature?: number;
}

const CHAT_COMPLETIONS_TIMEOUT_MS = 60_000;
const defaultConfidenceThreshold = 0.78;
const defaultMaxKeywords = 16;

export async function normalizeProductReviewKeywords(
  product: PdpProductSignal,
  locale: PdpGeoLocale,
  market: string | undefined,
  options: PdpGeoGeneratorOptions
): Promise<KeywordNormalizationApplication> {
  const evidence: PdpGeoEvidence[] = [];
  const warnings: string[] = [];
  let normalizedProduct = product;
  let usage: PdpGeoTokenUsage | undefined;

  const settings = options.keywordNormalization;
  const resolved = resolveKeywordNormalizer(options);
  if (!resolved.normalizer) {
    if (settings?.enabled && resolved.warning) {
      warnings.push(resolved.warning);
      evidence.push({ field: "reviews.keywords", source: "llm", value: resolved.warning });
    }
    return {
      product: normalizedProduct,
      evidence,
      warnings
    };
  }

  const request = createKeywordNormalizationRequest(normalizedProduct, locale, market, settings);
  if (request.reviewKeywords.length === 0) {
    return {
      product: normalizedProduct,
      evidence,
      warnings
    };
  }

  try {
    const result = await resolved.normalizer.normalizeKeywords(request);
    usage = result.usage;
    const accepted = selectSafeCorrections(
      request.reviewKeywords,
      result.corrections,
      locale,
      settings?.confidenceThreshold ?? defaultConfidenceThreshold
    );

    if (accepted.length > 0) {
      normalizedProduct = applyReviewKeywordCorrections(normalizedProduct, accepted);
      evidence.push({
        field: "reviews.keywords",
        source: "llm",
        value: `Accepted model-backed keyword corrections: ${formatCorrectionSummary(accepted)}`
      });
    }

    warnings.push(...(result.warnings ?? []));
    for (const warning of result.warnings ?? []) {
      evidence.push({ field: "reviews.keywords", source: "llm", value: warning });
    }
  } catch (error) {
    const message = error instanceof Error ? error.message : "Keyword normalization provider failed.";
    warnings.push(message);
    evidence.push({
      field: "reviews.keywords",
      source: "llm",
      value: `Keyword normalization skipped: ${message}`
    });
  }

  return {
    product: normalizedProduct,
    evidence,
    warnings,
    usage
  };
}

export class ModelBackedKeywordNormalizer implements PdpGeoKeywordNormalizer {
  constructor(private readonly config: ModelBackedKeywordNormalizerConfig) {}

  async normalizeKeywords(request: PdpGeoKeywordNormalizationRequest): Promise<PdpGeoKeywordNormalizationResult> {
    switch (this.config.provider) {
      case "openai":
        return this.normalizeWithOpenAI(request);
      case "gemini":
        return this.normalizeWithGemini(request);
      case "azure-openai":
        return this.normalizeWithAzureApi(request);
      case "custom":
      case "mock":
      default:
        return { corrections: [], warnings: [`${this.config.provider} keyword normalization provider has no model-backed adapter.`] };
    }
  }

  private async normalizeWithOpenAI(request: PdpGeoKeywordNormalizationRequest): Promise<PdpGeoKeywordNormalizationResult> {
    if (!this.config.apiKey || !this.config.model) {
      throw new Error("OPENAI_API_KEY and OPENAI_MODEL are required for keyword normalization.");
    }

    const prompt = createKeywordNormalizationPrompt(request);
    const response = await fetch("https://api.openai.com/v1/responses", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.config.apiKey}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        model: this.config.model,
        instructions: prompt.system,
        input: prompt.user
      })
    });

    if (!response.ok) {
      throw new Error(`OpenAI keyword normalization failed: ${response.status}`);
    }

    const payloadText = await response.text();
    const payload = JSON.parse(payloadText) as Record<string, unknown>;
    return {
      ...parseKeywordNormalizationJson(payloadText),
      usage: tokenUsageFromOpenAi(payload.usage)
    };
  }

  private async normalizeWithGemini(request: PdpGeoKeywordNormalizationRequest): Promise<PdpGeoKeywordNormalizationResult> {
    if (!this.config.apiKey || !this.config.model) {
      throw new Error("GEMINI_API_KEY and GEMINI_MODEL are required for keyword normalization.");
    }

    const prompt = createKeywordNormalizationPrompt(request);
    const response = await fetch(`https://generativelanguage.googleapis.com/v1beta/models/${this.config.model}:generateContent`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-goog-api-key": this.config.apiKey
      },
      body: JSON.stringify({
        systemInstruction: { parts: [{ text: prompt.system }] },
        contents: [{ role: "user", parts: [{ text: prompt.user }] }]
      })
    });

    if (!response.ok) {
      throw new Error(`Gemini keyword normalization failed: ${response.status}`);
    }

    const payload = await response.json() as {
      candidates?: Array<{ content?: { parts?: Array<{ text?: string }> } }>;
      usageMetadata?: unknown;
    };
    const rawText = payload.candidates?.[0]?.content?.parts?.map((part) => part.text ?? "").join("\n") ?? "";
    return {
      ...parseKeywordNormalizationJson(rawText),
      usage: tokenUsageFromGemini(payload.usageMetadata)
    };
  }

  private async normalizeWithAzureApi(request: PdpGeoKeywordNormalizationRequest): Promise<PdpGeoKeywordNormalizationResult> {
    if (!this.config.apiKey || !this.config.endpoint || !this.config.deployment) {
      throw new Error("AZURE_OPENAI_API_KEY, AZURE_OPENAI_ENDPOINT, and AZURE_OPENAI_DEPLOYMENT are required for keyword normalization.");
    }

    const apiVersion = this.config.apiVersion ?? "2025-04-01-preview";
    const endpoint = this.config.endpoint.replace(/\/$/, "");
    const url = `${endpoint}/openai/deployments/${this.config.deployment}/chat/completions?api-version=${apiVersion}`;
    const prompt = createKeywordNormalizationPrompt(request);
    const response = await fetchWithTimeout(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "api-key": this.config.apiKey
      },
      body: JSON.stringify({
        messages: [
          { role: "system", content: prompt.system },
          { role: "user", content: prompt.user }
        ],
        ...temperatureBody(this.config.temperature)
      })
    }, CHAT_COMPLETIONS_TIMEOUT_MS, "Azure keyword normalization");

    if (!response.ok) {
      throw new Error(`Azure keyword normalization failed: ${response.status}`);
    }

    const payload = await response.json() as { choices?: Array<{ message?: { content?: string } }>; usage?: unknown };
    return {
      ...parseKeywordNormalizationJson(payload.choices?.[0]?.message?.content ?? ""),
      usage: tokenUsageFromChatCompletions(payload.usage)
    };
  }
}

function tokenUsageFromOpenAi(value: unknown): PdpGeoTokenUsage | undefined {
  if (!value || typeof value !== "object") {
    return undefined;
  }
  const usage = value as Record<string, unknown>;
  return compactTokenUsage({
    inputTokens: numberField(usage.input_tokens) ?? numberField(usage.prompt_tokens),
    outputTokens: numberField(usage.output_tokens) ?? numberField(usage.completion_tokens),
    totalTokens: numberField(usage.total_tokens)
  });
}

function tokenUsageFromGemini(value: unknown): PdpGeoTokenUsage | undefined {
  if (!value || typeof value !== "object") {
    return undefined;
  }
  const usage = value as Record<string, unknown>;
  return compactTokenUsage({
    inputTokens: numberField(usage.promptTokenCount),
    outputTokens: numberField(usage.candidatesTokenCount),
    totalTokens: numberField(usage.totalTokenCount)
  });
}

function tokenUsageFromChatCompletions(value: unknown): PdpGeoTokenUsage | undefined {
  if (!value || typeof value !== "object") {
    return undefined;
  }
  const usage = value as Record<string, unknown>;
  const inputTokens = numberField(usage.prompt_tokens) ?? numberField(usage.input_tokens);
  const outputTokens = numberField(usage.completion_tokens) ?? numberField(usage.output_tokens);
  return compactTokenUsage({
    inputTokens,
    outputTokens,
    totalTokens: numberField(usage.total_tokens) ?? sumOptional(inputTokens, outputTokens)
  });
}

function compactTokenUsage(usage: PdpGeoTokenUsage): PdpGeoTokenUsage | undefined {
  return usage.inputTokens !== undefined || usage.outputTokens !== undefined || usage.totalTokens !== undefined ? usage : undefined;
}

function sumOptional(left: number | undefined, right: number | undefined): number | undefined {
  if (left === undefined && right === undefined) {
    return undefined;
  }
  return (left ?? 0) + (right ?? 0);
}

function numberField(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

async function fetchWithTimeout(url: string, init: RequestInit, timeoutMs: number, label: string): Promise<Response> {
  const { signal, cancel } = createTimeoutSignal(timeoutMs);

  try {
    return await fetch(url, {
      ...init,
      signal
    });
  } catch (error) {
    if (isAbortError(error)) {
      throw new Error(`${label} timed out after ${Math.round(timeoutMs / 1000)}s.`);
    }
    throw error;
  } finally {
    cancel();
  }
}

function createTimeoutSignal(timeoutMs: number): { signal: AbortSignal; cancel: () => void } {
  if (typeof AbortSignal.timeout === "function") {
    return {
      signal: AbortSignal.timeout(timeoutMs),
      cancel: () => {}
    };
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  return {
    signal: controller.signal,
    cancel: () => clearTimeout(timeout)
  };
}

function isAbortError(error: unknown): boolean {
  return typeof error === "object"
    && error !== null
    && "name" in error
    && (error as { name?: unknown }).name === "AbortError";
}

function resolveKeywordNormalizer(options: PdpGeoGeneratorOptions): { normalizer?: PdpGeoKeywordNormalizer; warning?: string } {
  if (options.customKeywordNormalizer) {
    return { normalizer: options.customKeywordNormalizer };
  }

  const settings = options.keywordNormalization;
  if (!settings?.enabled) {
    return {};
  }

  const provider = settings.provider ?? options.provider ?? "mock";
  if (provider === "mock" || provider === "custom") {
    return { warning: `${provider} keyword normalization requires customKeywordNormalizer.` };
  }

  return {
    normalizer: new ModelBackedKeywordNormalizer({
      provider,
      apiKey: settings.apiKey ?? options.apiKey,
      model: settings.model ?? options.model,
      endpoint: settings.endpoint ?? options.endpoint,
      deployment: settings.deployment ?? options.deployment,
      apiVersion: settings.apiVersion ?? options.apiVersion,
      temperature: options.temperature
    })
  };
}

function createKeywordNormalizationRequest(
  product: PdpProductSignal,
  locale: PdpGeoLocale,
  market: string | undefined,
  settings?: PdpGeoKeywordNormalizationSettings
): PdpGeoKeywordNormalizationRequest {
  const maxKeywords = settings?.maxKeywords ?? defaultMaxKeywords;
  return {
    productName: product.name,
    locale,
    market,
    reviewKeywords: unique(product.reviews.keywords).slice(0, maxKeywords),
    reviewBodies: unique(product.reviews.items.map((review) => review.body)).slice(0, 6),
    benefits: unique(product.benefits).slice(0, 8),
    effects: unique(product.effects).slice(0, 8),
    sourceTexts: unique(product.sourceTexts).slice(0, 10)
  };
}

function applyReviewKeywordCorrections(product: PdpProductSignal, corrections: PdpGeoKeywordCorrection[]): PdpProductSignal {
  if (corrections.length === 0) {
    return product;
  }
  const correctionMap = new Map(corrections.map((correction) => [keywordKey(correction.original), cleanKeyword(correction.normalized)]));
  return {
    ...product,
    reviews: {
      ...product.reviews,
      keywords: unique(product.reviews.keywords.map((keyword) => correctionMap.get(keywordKey(keyword)) ?? keyword))
    }
  };
}

function selectSafeCorrections(
  sourceKeywords: string[],
  corrections: PdpGeoKeywordCorrection[],
  locale: PdpGeoLocale,
  confidenceThreshold: number
): PdpGeoKeywordCorrection[] {
  const sourceMap = new Map(sourceKeywords.map((keyword) => [keywordKey(keyword), keyword]));
  const accepted: PdpGeoKeywordCorrection[] = [];

  for (const correction of corrections) {
    const original = sourceMap.get(keywordKey(correction.original));
    const normalized = cleanKeyword(correction.normalized);
    if (!original || !normalized || correction.confidence < confidenceThreshold) {
      continue;
    }
    if (keywordKey(original) === keywordKey(normalized) || !isSafeKeywordCorrection(original, normalized, locale)) {
      continue;
    }
    accepted.push({
      original,
      normalized,
      confidence: correction.confidence,
      reason: correction.reason
    });
  }

  return uniqueCorrections(accepted);
}

function isSafeKeywordCorrection(original: string, normalized: string, locale: PdpGeoLocale): boolean {
  const source = cleanKeyword(original);
  const target = cleanKeyword(normalized);
  if (!target || target.length < 2 || target.length > 32 || /[,/|;:()[\]{}<>]/.test(target) || /https?:\/\//i.test(target)) {
    return false;
  }
  if (target.split(/\s+/).length > 4) {
    return false;
  }
  if (hasMixedLanguageExpansion(source, target)) {
    return false;
  }

  const distance = levenshteinDistance(keywordFingerprint(source), keywordFingerprint(target));
  const limit = correctionDistanceLimit(source, target, locale);
  return distance <= limit;
}

function correctionDistanceLimit(original: string, normalized: string, locale: PdpGeoLocale): number {
  const length = Math.max(keywordFingerprint(original).length, keywordFingerprint(normalized).length);
  if (locale === "ko-KR" || /[가-힣]/.test(`${original}${normalized}`)) {
    return length <= 4 ? 2 : Math.max(2, Math.floor(length * 0.35));
  }
  return length <= 8 ? 2 : Math.max(2, Math.floor(length * 0.25));
}

function hasMixedLanguageExpansion(original: string, normalized: string): boolean {
  const sourceHasKorean = /[가-힣]/.test(original);
  const targetHasKorean = /[가-힣]/.test(normalized);
  const sourceHasLatin = /[A-Za-z]/.test(original);
  const targetHasLatin = /[A-Za-z]/.test(normalized);
  return (sourceHasKorean && targetHasLatin && !sourceHasLatin) || (sourceHasLatin && targetHasKorean && !sourceHasKorean);
}

function createKeywordNormalizationPrompt(request: PdpGeoKeywordNormalizationRequest): { system: string; user: string } {
  return {
    system: [
      "You are a conservative typo-normalization agent for product review keywords.",
      "Return strict JSON only: {\"corrections\":[{\"original\":\"\",\"normalized\":\"\",\"confidence\":0.0,\"reason\":\"\"}],\"warnings\":[]}.",
      "Only correct obvious typos, OCR mistakes, keyboard-adjacent mistakes, or spacing mistakes in the original keyword language.",
      "Do not translate, expand, summarize, add new claims, add new benefits, or rewrite a keyword into marketing copy.",
      "If uncertain, omit the correction. Keep normalized keywords concise and source-backed."
    ].join("\n"),
    user: JSON.stringify({
      task: "Normalize only misspelled review keywords. Leave valid keywords unchanged by omitting them from corrections.",
      productName: request.productName,
      locale: request.locale,
      market: request.market,
      reviewKeywords: request.reviewKeywords,
      context: {
        reviewBodies: request.reviewBodies,
        benefits: request.benefits,
        effects: request.effects,
        sourceTexts: request.sourceTexts
      }
    }, null, 2)
  };
}

function parseKeywordNormalizationJson(text: string): PdpGeoKeywordNormalizationResult {
  const rawText = parseProviderText(text);
  const jsonMatch = rawText.match(/\{[\s\S]*\}/);
  if (!jsonMatch) {
    return { corrections: [], warnings: ["No parseable keyword normalization JSON returned."], rawText };
  }

  const payload = JSON.parse(jsonMatch[0]) as Partial<PdpGeoKeywordNormalizationResult>;
  return {
    corrections: Array.isArray(payload.corrections)
      ? payload.corrections
        .map(normalizeCorrection)
        .filter((correction): correction is PdpGeoKeywordCorrection => Boolean(correction))
      : [],
    warnings: Array.isArray(payload.warnings) ? payload.warnings.map(String).filter(Boolean) : undefined,
    rawText
  };
}

function parseProviderText(text: string): string {
  try {
    const parsed = JSON.parse(text) as {
      output_text?: string;
      output?: Array<{ content?: Array<{ text?: string }> }>;
    };
    return parsed.output_text
      ?? parsed.output?.flatMap((item) => item.content ?? []).map((item) => item.text ?? "").join("\n")
      ?? text;
  } catch {
    return text;
  }
}

function normalizeCorrection(value: unknown): PdpGeoKeywordCorrection | undefined {
  if (!isRecord(value)) {
    return undefined;
  }
  const original = typeof value.original === "string" ? cleanKeyword(value.original) : "";
  const normalized = typeof value.normalized === "string" ? cleanKeyword(value.normalized) : "";
  const confidence = typeof value.confidence === "number" ? value.confidence : Number(value.confidence);
  if (!original || !normalized || !Number.isFinite(confidence)) {
    return undefined;
  }
  return {
    original,
    normalized,
    confidence,
    reason: typeof value.reason === "string" ? value.reason : undefined
  };
}

function formatCorrectionSummary(corrections: PdpGeoKeywordCorrection[]): string {
  return corrections.map((correction) => `${correction.original} -> ${correction.normalized}`).join(", ");
}

function uniqueCorrections(corrections: PdpGeoKeywordCorrection[]): PdpGeoKeywordCorrection[] {
  const seen = new Set<string>();
  return corrections.filter((correction) => {
    const key = `${keywordKey(correction.original)}:${keywordKey(correction.normalized)}`;
    if (seen.has(key)) {
      return false;
    }
    seen.add(key);
    return true;
  });
}

function keywordFingerprint(value: string): string {
  return cleanKeyword(value).toLowerCase().replace(/[^a-z0-9가-힣ぁ-んァ-ン一-龯]+/gi, "");
}

function keywordKey(value: string): string {
  return cleanKeyword(value).toLowerCase();
}

function cleanKeyword(value: string): string {
  return value.replace(/\s+/g, " ").replace(/\s+([,.!?。！？])/g, "$1").trim();
}

function unique(values: string[]): string[] {
  return Array.from(new Set(values.map(cleanKeyword).filter(Boolean)));
}

function levenshteinDistance(left: string, right: string): number {
  if (left === right) {
    return 0;
  }
  if (!left) {
    return right.length;
  }
  if (!right) {
    return left.length;
  }

  const previous = Array.from({ length: right.length + 1 }, (_, index) => index);
  const current = Array.from({ length: right.length + 1 }, () => 0);

  for (let leftIndex = 1; leftIndex <= left.length; leftIndex += 1) {
    current[0] = leftIndex;
    for (let rightIndex = 1; rightIndex <= right.length; rightIndex += 1) {
      const cost = left[leftIndex - 1] === right[rightIndex - 1] ? 0 : 1;
      const insertion = (current[rightIndex - 1] ?? 0) + 1;
      const deletion = (previous[rightIndex] ?? 0) + 1;
      const substitution = (previous[rightIndex - 1] ?? 0) + cost;
      current[rightIndex] = Math.min(
        insertion,
        deletion,
        substitution
      );
    }
    for (let index = 0; index < previous.length; index += 1) {
      previous[index] = current[index] ?? 0;
    }
  }

  return previous[right.length] ?? 0;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: packages/pdp-geo-generator-agent/src/pdp_geo_generator_agent/keyword_normalizer.py
"""Optional review-keyword typo normalization with conservative correction gates."""

from __future__ import annotations

import inspect
import math
import re
from collections.abc import Mapping, Sequence
from typing import Any, TypedDict

from neo_js_compat import js_json_pretty_dumps

from ._json import as_dict, as_list, clean_text

_DEFAULT_CONFIDENCE_THRESHOLD = 0.78
_DEFAULT_MAX_KEYWORDS = 16


class _KeywordNormalizationPrompt(TypedDict):
    system: str
    user: str


_KEYWORD_NORMALIZATION_SYSTEM_PROMPT = "\n".join(
    (
        "You are a conservative typo-normalization agent for product review keywords.",
        'Return strict JSON only: {"corrections":[{"original":"","normalized":"","confidence":0.0,"reason":""}],"warnings":[]}.',
        "Only correct obvious typos, OCR mistakes, keyboard-adjacent mistakes, or spacing mistakes in the original keyword language.",
        "Do not translate, expand, summarize, add new claims, add new benefits, or rewrite a keyword into marketing copy.",
        "If uncertain, omit the correction. Keep normalized keywords concise and source-backed.",
    )
)


async def normalize_product_review_keywords(
    product: Mapping[str, Any], locale: str, market: str | None, options: Mapping[str, Any] | None = None
) -> dict[str, Any]:
    """Normalize only high-confidence spelling-level review keyword corrections."""

    runtime = as_dict(options)
    evidence: list[dict[str, str]] = []
    warnings: list[str] = []
    normalized_product = _clone_product(product)
    normalizer, resolution_warning = _resolve_keyword_normalizer(runtime)
    settings = as_dict(runtime.get("keywordNormalization"))
    if normalizer is None:
        if settings.get("enabled") is True and resolution_warning:
            warnings.append(resolution_warning)
            evidence.append({"field": "reviews.keywords", "source": "llm", "value": resolution_warning})
        return {"product": normalized_product, "evidence": evidence, "warnings": warnings, "called": False, "applied": False}

    request = _create_request(normalized_product, locale, market, settings)
    if not request["reviewKeywords"]:
        return {"product": normalized_product, "evidence": evidence, "warnings": warnings, "called": False, "applied": False}
    try:
        result = await _call_keyword_normalizer(normalizer, request)
        accepted = select_safe_keyword_corrections(
            request["reviewKeywords"],
            [as_dict(item) for item in as_list(result.get("corrections"))],
            locale,
            _finite_number(settings.get("confidenceThreshold"), _DEFAULT_CONFIDENCE_THRESHOLD),
        )
        if accepted:
            normalized_product = _apply_corrections(normalized_product, accepted)
            evidence.append(
                {
                    "field": "reviews.keywords",
                    "source": "llm",
                    "value": "Accepted model-backed keyword corrections: "
                    + ", ".join(f"{item['original']} -> {item['normalized']}" for item in accepted),
                }
            )
        warnings.extend(clean_text(item) for item in as_list(result.get("warnings")) if clean_text(item))
        evidence.extend({"field": "reviews.keywords", "source": "llm", "value": warning} for warning in warnings)
        return {
            "product": normalized_product,
            "evidence": evidence,
            "warnings": _unique(warnings),
            "called": True,
            "applied": bool(accepted),
            **({"usage": result["usage"]} if result.get("usage") is not None else {}),
        }
    except Exception as error:  # a keyword typo model must never make generation fail
        message = str(error) or "Keyword normalization provider failed."
        warnings.append(message)
        evidence.append(
            {"field": "reviews.keywords", "source": "llm", "value": f"Keyword normalization skipped: {message}"}
        )
        return {"product": normalized_product, "evidence": evidence, "warnings": warnings, "called": True, "applied": False}


class ModelBackedKeywordNormalizer:
    """Provider-neutral bridge for the optional keyword-normalization stage."""

    def __init__(self, config: Mapping[str, Any]) -> None:
        self.config = dict(config)

    async def normalize_keywords(self, request: Mapping[str, Any]) -> Mapping[str, Any]:
        provider_name = clean_text(self.config.get("provider"))
        if provider_name in {"aistudio", "custom", "mock"}:
            return {
                "corrections": [],
                "warnings": [f"{provider_name} keyword normalization provider has no model-backed adapter."],
            }
        from .providers import create_provider

        provider = create_provider(self.config)
        prompt = create_keyword_normalization_prompt(request)
        return await provider.generate_json(
            stage="keyword-normalization",
            system=prompt["system"],
            user=prompt["user"],
        )

    async def normalizeKeywords(self, request: Mapping[str, Any]) -> Mapping[str, Any]:
        return await self.normalize_keywords(request)


def create_keyword_normalization_prompt(request: Mapping[str, Any]) -> _KeywordNormalizationPrompt:
    """Build the source-selected conservative keyword-normalization prompt."""

    context: dict[str, Any] = {
        field: request[field]
        for field in ("reviewBodies", "benefits", "effects", "sourceTexts")
        if field in request
    }
    payload: dict[str, Any] = {
        "task": "Normalize only misspelled review keywords. Leave valid keywords unchanged by omitting them from corrections.",
        **{field: request[field] for field in ("productName", "locale", "market", "reviewKeywords") if field in request},
        "context": context,
    }
    return {
        "system": _KEYWORD_NORMALIZATION_SYSTEM_PROMPT,
        "user": js_json_pretty_dumps(payload),
    }


PDP_GEO_KEYWORD_NORMALIZATION_JSON_SCHEMA: dict[str, Any] = {
    "type": "object",
    "additionalProperties": False,
    "properties": {
        "corrections": {
            "type": "array",
            "items": {
                "type": "object",
                "additionalProperties": False,
                "properties": {
                    "original": {"type": "string"},
                    "normalized": {"type": "string"},
                    "confidence": {"type": "number"},
                    "reason": {"type": "string"},
                },
            },
        },
        "warnings": {"type": "array", "items": {"type": "string"}},
    },
}


def select_safe_keyword_corrections(
    source_keywords: Sequence[str], corrections: Sequence[Mapping[str, Any]], locale: str, confidence_threshold: float
) -> list[dict[str, Any]]:
    source = {_key(value): value for value in source_keywords if _clean_keyword(value)}
    accepted: list[dict[str, Any]] = []
    for correction in corrections:
        original = source.get(_key(clean_text(correction.get("original"))))
        normalized = _clean_keyword(clean_text(correction.get("normalized")))
        confidence = _finite_number(correction.get("confidence"), -1)
        if not original or not normalized or confidence < confidence_threshold:
            continue
        if _key(original) == _key(normalized) or not _is_safe_correction(original, normalized, locale):
            continue
        accepted.append(
            {
                "original": original,
                "normalized": normalized,
                "confidence": confidence,
                **({"reason": clean_text(correction.get("reason"))} if clean_text(correction.get("reason")) else {}),
            }
        )
    result: list[dict[str, Any]] = []
    seen: set[str] = set()
    for correction in accepted:
        key = f"{_key(correction['original'])}:{_key(correction['normalized'])}"
        if key not in seen:
            seen.add(key)
            result.append(correction)
    return result


async def _call_keyword_normalizer(normalizer: object, request: Mapping[str, Any]) -> dict[str, Any]:
    method = getattr(normalizer, "normalize_keywords", None) or getattr(normalizer, "normalizeKeywords", None)
    if not callable(method):
        raise TypeError(
            "customKeywordNormalizer must provide normalize_keywords(request) or normalizeKeywords(request)."
        )
    result = method(dict(request))
    if inspect.isawaitable(result):
        result = await result
    return as_dict(result)


def _resolve_keyword_normalizer(runtime: Mapping[str, Any]) -> tuple[object | None, str | None]:
    custom = runtime.get("customKeywordNormalizer")
    if custom is not None:
        return custom, None
    settings = as_dict(runtime.get("keywordNormalization"))
    if settings.get("enabled") is not True:
        return None, None
    provider = clean_text(settings.get("provider") or runtime.get("provider") or "mock")
    if provider in {"mock", "custom"}:
        return None, f"{provider} keyword normalization requires customKeywordNormalizer."
    # The retained implementation has no AI Studio keyword adapter.  Keep this
    # explicit rather than silently posting a chat-completions request under a
    # stage that TypeScript would report as unavailable.
    if provider == "aistudio":
        return None, "aistudio keyword normalization provider has no model-backed adapter."
    inherits = not settings.get("provider") or settings.get("provider") == runtime.get("provider")
    return (
        ModelBackedKeywordNormalizer(
            {
                "provider": provider,
                "apiKey": settings.get("apiKey") or (runtime.get("apiKey") if inherits else None),
                "model": settings.get("model") or (runtime.get("model") if inherits else None),
                "endpoint": settings.get("endpoint") or (runtime.get("endpoint") if inherits else None),
                "deployment": settings.get("deployment") or (runtime.get("deployment") if inherits else None),
                "apiVersion": settings.get("apiVersion") or (runtime.get("apiVersion") if inherits else None),
                "temperature": runtime.get("temperature"),
            }
        ),
        None,
    )


def _create_request(
    product: Mapping[str, Any], locale: str, market: str | None, settings: Mapping[str, Any]
) -> dict[str, Any]:
    reviews = as_dict(product.get("reviews"))
    max_keywords = max(0, _integer(settings.get("maxKeywords"), _DEFAULT_MAX_KEYWORDS))
    return {
        "productName": clean_text(product.get("name")),
        "locale": locale,
        **({"market": market} if market else {}),
        "reviewKeywords": _unique(clean_text(item) for item in as_list(reviews.get("keywords")))[:max_keywords],
        "reviewBodies": _unique(clean_text(as_dict(item).get("body")) for item in as_list(reviews.get("items")))[:6],
        "benefits": _unique(clean_text(item) for item in as_list(product.get("benefits")))[:8],
        "effects": _unique(clean_text(item) for item in as_list(product.get("effects")))[:8],
        "sourceTexts": _unique(clean_text(item) for item in as_list(product.get("sourceTexts")))[:10],
    }


def _apply_corrections(product: Mapping[str, Any], corrections: Sequence[Mapping[str, Any]]) -> dict[str, Any]:
    result = dict(product)
    reviews = dict(as_dict(product.get("reviews")))
    changes = {_key(clean_text(item.get("original"))): clean_text(item.get("normalized")) for item in corrections}
    reviews["keywords"] = _unique(
        changes.get(_key(clean_text(keyword)), clean_text(keyword)) for keyword in as_list(reviews.get("keywords"))
    )
    result["reviews"] = reviews
    return result


def _is_safe_correction(original: str, normalized: str, locale: str) -> bool:
    if (
        len(normalized) < 2
        or len(normalized) > 32
        or re.search(r"[,/|;:()[\]{}<>]", normalized)
        or re.search(r"https?://", normalized, re.I)
        or len(normalized.split()) > 4
        or _mixed_language_expansion(original, normalized)
    ):
        return False
    source, target = _fingerprint(original), _fingerprint(normalized)
    return _levenshtein(source, target) <= _distance_limit(original, normalized, locale)


def _distance_limit(original: str, normalized: str, locale: str) -> int:
    length = max(len(_fingerprint(original)), len(_fingerprint(normalized)))
    if locale == "ko-KR" or re.search(r"[가-힣]", f"{original}{normalized}"):
        return 2 if length <= 4 else max(2, int(length * 0.35))
    return 2 if length <= 8 else max(2, int(length * 0.25))


def _mixed_language_expansion(original: str, normalized: str) -> bool:
    source_korean, target_korean = bool(re.search(r"[가-힣]", original)), bool(re.search(r"[가-힣]", normalized))
    source_latin, target_latin = bool(re.search(r"[A-Za-z]", original)), bool(re.search(r"[A-Za-z]", normalized))
    return (source_korean and target_latin and not source_latin) or (
        source_latin and target_korean and not source_korean
    )


def _fingerprint(value: str) -> str:
    return re.sub(r"[^a-z0-9가-힣ぁ-んァ-ン一-龯]+", "", _clean_keyword(value).casefold())


def _key(value: str) -> str:
    return _clean_keyword(value).casefold()


def _clean_keyword(value: str) -> str:
    return re.sub(r"\s+([,.!?。！？])", r"\1", re.sub(r"\s+", " ", value)).strip()


def _levenshtein(left: str, right: str) -> int:
    if left == right:
        return 0
    if not left:
        return len(right)
    if not right:
        return len(left)
    previous = list(range(len(right) + 1))
    for left_index, left_character in enumerate(left, 1):
        current = [left_index]
        for right_index, right_character in enumerate(right, 1):
            current.append(
                min(
                    current[-1] + 1,
                    previous[right_index] + 1,
                    previous[right_index - 1] + (left_character != right_character),
                )
            )
        previous = current
    return previous[-1]


def _clone_product(product: Mapping[str, Any]) -> dict[str, Any]:
    result = dict(product)
    if isinstance(result.get("reviews"), Mapping):
        reviews = dict(as_dict(result["reviews"]))
        reviews["keywords"] = list(as_list(reviews.get("keywords")))
        reviews["items"] = [dict(as_dict(item)) for item in as_list(reviews.get("items"))]
        result["reviews"] = reviews
    return result


def _finite_number(value: object, default: float) -> float:
    if isinstance(value, bool):
        return default
    if isinstance(value, int | float | str):
        try:
            number = float(value)
        except ValueError:
            return default
        return number if math.isfinite(number) else default
    return default


def _integer(value: object, default: int) -> int:
    return int(value) if isinstance(value, int | float) and not isinstance(value, bool) else default


def _unique(values: Sequence[str] | Any) -> list[str]:
    result: list[str] = []
    seen: set[str] = set()
    for raw in values:
        value = _clean_keyword(clean_text(raw))
        if value and value.casefold() not in seen:
            seen.add(value.casefold())
            result.append(value)
    return result


normalizeProductReviewKeywords = normalize_product_review_keywords
selectSafeKeywordCorrections = select_safe_keyword_corrections

__all__ = [
    "ModelBackedKeywordNormalizer",
    "PDP_GEO_KEYWORD_NORMALIZATION_JSON_SCHEMA",
    "normalize_product_review_keywords",
    "normalizeProductReviewKeywords",
    "select_safe_keyword_corrections",
    "selectSafeKeywordCorrections",
]
